Move internal diophantine solvers into classes

ethankward opened the pull request:
#### References to other Issues or PRs
#18493

#### Brief description of what is fixed or changed
Move all internal solvers (`_diop_linear`, `_diop_quadratic`, `_diop_ternary_quadratic`, `_diop_ternary_quadratic_normal`, `_diop_general_pythagorean`, `_diop_general_sum_of_squares`, `_diop_general_sum_of_even_powers`) into a new `solve` method in `DiophantineEquationType`.

Also added `diop_solve` to `__init__.py`.

Added a new parameter `n_parameters` to `DiophantineSolutionSet` to explicitly control the number of parameters.

#### Other comments
`_diop_ternary_quadratic` and `_diop_ternary_quadratic_normal` have been retained as placeholders for now as they are used in other functions.

#### Release Notes
<!-- BEGIN RELEASE NOTES -->
NO ENTRY
<!-- END RELEASE NOTES -->

ethankward committed:
move internal solvers into DiophantineEquationType classes

ethankward committed:
add diop_solve

ethankward committed:
update test_diop_general_sum_of_squares_quick to not test private function

oscarbenjamin commented on sympy/solvers/diophantine/diophantine.py:
```diff
@@ -190,27 +188,391 @@ def matches(self):
 
+    def solve(self, params=None, limit=None):
+        """
+        Solves diophantine equations of the form:
```
Would this make more sense as a class-level docstring?

ethankward replied:
Yes. I will merge this with the docstring that is currently in the `diop_linear` function.

oscarbenjamin commented:
Are you done with this?

ethankward commented:
No, I'm going to make documentation changes as you suggested.

ethankward committed:
Add n_parameters parameter to DiophantineSolutionSet. Add docstrings to DiophantineEquationType classes. Check for match before attempting to solve. Update GeneralPythagorean parameter handling.

ethankward committed:
Test GeneralPythagorean. Additional tests in test_diophantine_solution_set for n_parameters.

ethankward commented:
OK, I had to make some more changes as I found some bugs. This is ready for review again.

ethankward committed:
InhomogeneousTernaryQuadratic docstring

oscarbenjamin commented:
Do the new docstrings need to be added to the rst docs somewhere?

ethankward committed:
test test_quadratic_parameter_passing; update tests for check_param

ethankward committed:
update DiophantineSolutionSet call to avoid exhausting zip object; update check_param

ethankward commented:
I would like to avoid modifying the .rst files until this is in a final state, since various things will probably be deprecated.

oscarbenjamin commented on sympy/solvers/diophantine/diophantine.py:
```diff
@@ -98,19 +99,34 @@ class DiophantineSolutionSet(set):
             self.parameters = tuple(parameters)
 
+            if n_parameters is not None:
+                if len(self.parameters) < self.n_parameters:
```
All the logic above with `n_parameters` and `self.n_parameters` is confusing. Is there a need for `n_parameters` to be anything other than len(self.parameters)

ethankward replied:
I'll see if I can simplify this. If `parameters` is an optional argument then it's necessary to be able to control its length.

ethankward replied:
Ok, I have removed `n_parameters` from `DiophantineSolutionSet` and made `parameters` a required argument. Subsequently, I added a `get_parameters` method to the `DiophantineEquationType` that can be overridden in the individual solvers. This greatly simplifies the logic of dealing with the interaction of `n_parameters` / `parameters`.

oscarbenjamin replied:
That seems better

ethankward committed:
remove n_parameters tests; update params -> parameters

ethankward committed:
additional BinaryQuadratic test

ethankward committed:
remove n_parameters from DiophantineSolutionSet, make parameters a required argument; update solvers

ethankward commented:
This PR needs more work but I'm leaving for several days- I will return to this next week.

ethankward committed:
use n_parameters and parameters class properties; use pre_solve method

ethankward commented:
I moved `parameters` and `n_parameters` to class properties (so that parameters would not be recreated every time they are accessed). Also moved pre-solution checking to its own method.

ethankward committed:
linear parameter handling

oscarbenjamin commented:
Should the classes here be added to the rst docs somewhere so that the docstrings show up?

ethankward committed:
internal classes

oscarbenjamin commented:
This looks good. Are you done with it?

oscarbenjamin commented:
Is this finished?

czgdp1807 commented:
@ethankward @oscarbenjamin Can we merge this?

czgdp1807 commented on sympy/solvers/diophantine/diophantine.py:
```diff
@@ -98,18 +97,17 @@ class DiophantineSolutionSet(set):
     """
 
-    def __init__(self, symbols_seq, parameters=None):
+    def __init__(self, symbols_seq, parameters):
```
IMO, this is a public API change. Do we need a deprecation warning for this?

czgdp1807 replied:
@oscarbenjamin What do you say about this change and the overall PR?

oscarbenjamin replied:
I think the PR looks good. We should resolve the merge conflicts and get it in.

oscarbenjamin committed:
Merge branch 'master' into tmp_diophantine_merge

oscarbenjamin committed:
quality(diophantine): remove unneeded imports in docstrings

oscarbenjamin commented:
Thanks @ethankward for this. I think probably you were intending to do more with this but what is here so far is still good work and it's best to get it in so that others can contribute further.

I'll merge this now.
